### PR TITLE
Avoid duplicate Ids for split & merge

### PR DESCRIPTION
**Pull request by marvin21**
Improve the splitting functionality by avoiding the occurrence of duplicate ids.
Avoid multiple appending of Targetlabel to the id & name of
a cloned NodeTemplate.

Signed-off-by: Marvin Wohlfarth <st112462@stud.uni-stuttgart.de>

- [x ] Ensure that the commit message is [a good commit message](https://github.com/erlang/otp/wiki/Writing-good-commit-messages)
- [ ] Change in CHANGELOG.md described
- [ ] Tests created for changes
- [ ] Screenshots added (for UI changes)

### CODE REVIEW

**Comment by saatkamp** [comment]
LGTM

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 1]
```diff
@@ -1073,4 +1078,75 @@ public static void deriveWPD(TEntityType ci, List<String> errors) {
 		}
 		return res;
 	}
+
+	/**
+	 * Check the id's for new nodes, avoid repeating of target labels
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java]: The general rule is that you should not use an apostrophe to form the plurals of nouns -- https://en.oxforddictionaries.com/punctuation/apostrophe

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 2]
```diff
@@ -1073,4 +1078,75 @@ public static void deriveWPD(TEntityType ci, List<String> errors) {
...
+	 * @param duplicatedNode
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java]: Remove the empty parameter descriptions

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 3]
```diff
@@ -1073,4 +1078,75 @@ public static void deriveWPD(TEntityType ci, List<String> errors) {
...
+		} else {
+			duplicatedNode.setId(Util.makeNCName(tNodeTemplate.getId() + "-" + targetLabel));
+			duplicatedNode.setName(Util.makeNCName(tNodeTemplate.getName() + "-" + targetLabel));
+			return duplicatedNode;
+		}
+	}
+
+	/**
+	 * Fix the Ids to avoid duplicate Ids in capabilities
+	 * @param nodeTemplate
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java]: Remove empty parameter description

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 4]
```diff
@@ -1073,4 +1078,75 @@ public static void deriveWPD(TEntityType ci, List<String> errors) {
...
+	 * @return fixedCapabilities with fixed Ids
+	 */
+	public static TNodeTemplate.Capabilities cloneCapabilities(TNodeTemplate nodeTemplate) {
+		if (nodeTemplate.getCapabilities() == null) {
+			LOGGER.info("Capabilites are null, not cloned any capabilities for " +  nodeTemplate.getId());
+			return null;
+		}
+			TNodeTemplate.Capabilities fixedCapabilities = new TNodeTemplate.Capabilities();
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java]: Please fix ident

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 5]
```diff
@@ -1073,4 +1078,75 @@ public static void deriveWPD(TEntityType ci, List<String> errors) {
...
+	}
+
+	/**
+	 * Fix the Ids to avoid duplicate Ids in capabilities
+	 * @param nodeTemplate
+	 * @return fixedCapabilities with fixed Ids
+	 */
+	public static TNodeTemplate.Capabilities cloneCapabilities(TNodeTemplate nodeTemplate) {
+		if (nodeTemplate.getCapabilities() == null) {
+			LOGGER.info("Capabilites are null, not cloned any capabilities for " +  nodeTemplate.getId());
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java]: This should be DEBUG. It is good that you use the logger, but it should use the right log level. This is not a user information , but more a DEBUG information, isn't it?

DEBUG is also logged as default (currently) - https://github.com/eclipse/winery/blob/master/org.eclipse.winery.repository/src/main/resources/logback.xml

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java [hunk 6]
```diff
@@ -1073,4 +1078,75 @@ public static void deriveWPD(TEntityType ci, List<String> errors) {
...
+			duplicatedNode.setId(Util.makeNCName(tNodeTemplate.getId() + "-" + targetLabel));
+			duplicatedNode.setName(Util.makeNCName(tNodeTemplate.getName() + "-" + targetLabel));
+			return duplicatedNode;
+		}
+	}
+
+	/**
+	 * Fix the Ids to avoid duplicate Ids in capabilities
+	 * @param nodeTemplate
+	 * @return fixedCapabilities with fixed Ids
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/backend/BackendUtils.java]: Please also state that `null` is returned (in which case?)

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java [hunk 7]
```diff
@@ -41,6 +42,9 @@
 
 	private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(Splitting.class);
 
+	private static int newRelationshipIdCounter = 100;
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java]: Why is 100 chosen? Please add a comment.

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java [hunk 8]
```diff
@@ -185,8 +189,9 @@ public TTopologyTemplate split(TTopologyTemplate topologyTemplate) throws Splitt
 					// Otherwise, duplicate the considered node for each target label
 					for (String targetLabel: predecessorsTargetLabel) {
 						TNodeTemplate duplicatedNode = BackendUtils.clone(currentNode);
-						duplicatedNode.setId(Util.makeNCName(currentNode.getId() + "-" + targetLabel));
-						duplicatedNode.setName(Util.makeNCName(currentNode.getName() + "-" + targetLabel));
+						duplicatedNode = BackendUtils.checkId(duplicatedNode, currentNode, targetLabel);
+						//duplicatedNode.setId(Util.makeNCName(currentNode.getId() + "-" + targetLabel));
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java]: Please remove disabled code.

File: org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java [hunk 9]
```diff
@@ -400,6 +405,8 @@ public TTopologyTemplate matchingWithDefaultHostSelection (TTopologyTemplate top
 	 * @return modified topology with the replaced Node Templates
 	 */
 	public TTopologyTemplate injectNodeTemplates (TTopologyTemplate topologyTemplate, Map<String, TNodeTemplate> injectNodes) {
+		String id = UUID.randomUUID().toString();
```
Comment (koppor) [inline org.eclipse.winery.repository/src/main/java/org/eclipse/winery/repository/splitting/Splitting.java]: Please comment, why you chose the approach of a random UUID (and no counter, which you did above).

**Pull request merged**